docs/API.md

```diff
@@ -1,681 +1,1085 @@
+
+* The `"jsonrpc"` property is not checked.
+* The error code -32603 is used for invalid arguments, as opposed to -32602.
+* When standard error codes are used, the accompanying messages are not the standard ones.
+
+
+### Notes on the JSON-RPC 2.0 implementation
+
+* Extra top-level properties in request objects are allowed but ignored.
+* Extra properties in the `"params"` object are allowed for some methods but ignored for others.
+* Error messages from the API should be considered sensitive sas they leak details about the internals of the application and the system.
+* The error code -32601 is used when the `"method"` property is missing, as opposed to -32600.
+
+
+## Request handling
+
+When a method expects a string argument but receives an array or an object,
+the value may be interpreted as something like `"ARRAY(0x1a2d1d0)"` or `"HASH(0x1a2d2c8)"`.
+
+When a method expects a boolean argument, any kind of value is accepted.
+A number of values are interpreted as false: `false`, `null`, `""`, `"0"` and any number equal to zero.
+Everything else is interpreted as true.
+
+When a method expects an integer arguments, numbers encoded in strings are also accepted and used transparently,
+and numbers with fractions are rounded to the nearest integer.
+
+For details on when a *test* are performed after it's been requested,
+see the [architecture documentation](Architecture.md).
+
+
+## Error reporting
+
+If the request object is invalid JSON, an error with code `-32700` is reported.
+
+If no method is specified or an invalid method is specified, an error with code `-32601` is reported.
+
+All error states that occur after the RPC method has been identified are reported as internal errors with code `-32603`.
+
+
+## Privilege levels
+
+This API provides three classes of methods:
+
+* *Unrestricted* methods are available to anyone with access to the API.
+* *Authenticated* methods have parameters for username and API key credentials.
+* *Administrative* methods require that the connection to the API is opened from localhost (`127.0.0.1` or `::1`).
+
+
+## Data types
+
+This sections describes a number of data types used in this API. Each data type
+is based on a JSON data type, but additionally imposes its own restrictions.
+
+
+### Batch id
+
+Basic data type: number
+
+An positive integer.
+
+The unique id of a *batch*.
+
+
+### Domain name
+
+Basic data type: string
+
+1. If the string contains characters outside the ASCII character set,
+   it must be possible to convert the entire string to the equivalent IDN A-label.
+
+2. If the string is a single character, that character must be `.`.
+
+3. The length of the string must not be greater than 254 characters.
+
+4. When the string is split at `.` characters (after IDNA conversion,
+   if needed), each component part must be at most 63 characters long.
+
+5. Each such component part must also consist only of the characters
+   `0` to `9`, `a` to `z`, `A` to `Z` and `-`.
```

Comment:
If we have the classical limitation (as BASIC00 requires) then the label may not start nor end with a "-".

But as it is implemented, there are no visible limitations on the character set. The domain can contains any character. This is in conflict with BASIC00.